extensions/stats/plugin.cc

```diff
@@ -49,42 +50,52 @@ void clearTcpMetrics(::Wasm::Common::RequestInfo& request_info) {
   request_info.tcp_sent_bytes = 0;
   request_info.tcp_received_bytes = 0;
 }
+
 }  // namespace
 
-bool PluginRootContext::onConfigure(size_t) {
-  std::unique_ptr<WasmData> configuration = getConfiguration();
-  // Parse configuration JSON string.
-  JsonParseOptions json_options;
-  Status status =
-      JsonStringToMessage(configuration->toString(), &config_, json_options);
-  if (status != Status::OK) {
-    LOG_WARN(absl::StrCat("Cannot parse plugin configuration JSON string ",
-                          configuration->toString()));
-    return false;
-  }
+void PluginRootContext::initializeDimensions() {
+  // Clean-up existing expressions.
+  // Potential perf optimization: re-use the existing expressions.
+  cleanupExpressions();
 
-  status = ::Wasm::Common::extractLocalNodeMetadata(&local_node_info_);
-  if (status != Status::OK) {
-    LOG_WARN("cannot parse local node metadata ");
-    return false;
-  }
-  outbound_ = ::Wasm::Common::TrafficDirection::Outbound ==
-              ::Wasm::Common::getTrafficDirection();
+  // Seed the common metric tags with the default set.
+  tags_ = IstioDimensions::defaultTags();
 
-  // Local data does not change, so populate it on config load.
-  istio_dimensions_.init(outbound_, local_node_info_);
+  // Process the dimension overrides
+  for (const auto& metric : config_.metrics()) {
+    if (metric.dimensions().empty()) {
+      continue;
+    }
 
-  if (outbound_) {
-    peer_metadata_id_key_ = ::Wasm::Common::kUpstreamMetadataIdKey;
-    peer_metadata_key_ = ::Wasm::Common::kUpstreamMetadataKey;
-  } else {
-    peer_metadata_id_key_ = ::Wasm::Common::kDownstreamMetadataIdKey;
-    peer_metadata_key_ = ::Wasm::Common::kDownstreamMetadataKey;
+    // sort map keys
+    std::vector<std::string> keys;
```

Comment:
Why sort? The only reason I can think of is to make hash of istio dimension the same across configuration update. But `metrics_` map lifetime does not expand configuration reload...

Reply:
If we don't sort, then because proto map iteration is not fixed, we get two types of metrics:
a;=;value1;.;b;=;value2 and b;=;value2;.;a;=;value1. After regexes are applied, the two become identical, but unfortunately, they are not merged, so prom metrics is simply duplicated.

Reply:
Oh so order matters when hash IstioDimension?

Reply:
Order matter for the resolveWithFields. And for hashing, too.